zone/client.cpp

```diff
@@ -10129,3 +10137,437 @@ void Client::SetAFK(uint8 afk_flag) {
+						}
+					}
+					else if (IsKnockBackExempted())
+					{
+						//still potential to trigger this if you're knocked back off a
+						//HUGE fall that takes > 2.5 seconds
+						if (speed > 30.0f)
+						{
+							CheatDetected(MQWarpKnockBack, GetX(), GetY(), GetZ());
+						}
+					}
+					else if (!IsPortExempted())
+					{
+						if (!IsMQExemptedArea(zone->GetZoneID(), GetX(), GetY(), GetZ()))
+						{
+							if (speed > (runs * 2 * RuleR(Zone, MQWarpDetectionDistanceFactor)))
+							{
+								m_TimeSinceLastPositionCheck = cur_time;
+								m_DistanceSinceLastPositionCheck = 0.0f;
+								CheatDetected(MQWarp, GetX(), GetY(), GetZ());
+								//Death(this, 10000000, SPELL_UNKNOWN, _1H_BLUNT);
+							}
+							else
+							{
+								CheatDetected(MQWarpLight, GetX(), GetY(), GetZ());
+							}
+						}
+					}
+				}
+			}
+		}
+		m_TimeSinceLastPositionCheck = cur_time;
+		m_DistanceSinceLastPositionCheck = 0.0f;
+	}
+	m_KnockBackExemption = v;
+}
+
+void Client::SetPortExemption(bool v)
+{
+	if (v == true)
+	{
+		uint32 cur_time = Timer::GetCurrentTime();
+		if ((cur_time - m_TimeSinceLastPositionCheck) > 1000)
+		{
+			float speed = (m_DistanceSinceLastPositionCheck * 100) / (float)(cur_time - m_TimeSinceLastPositionCheck);
+			int runs = GetRunspeed();
+			if (speed > (runs * RuleR(Zone, MQWarpDetectionDistanceFactor)))
+			{
+				if (!GetGMSpeed() && (runs >= GetBaseRunspeed() || (speed > (GetBaseRunspeed() * RuleR(Zone, MQWarpDetectionDistanceFactor)))))
+				{
+					printf("%s %i moving too fast! moved: %.2f in %ims, speed %.2f\n", __FILE__, __LINE__,
+						m_DistanceSinceLastPositionCheck, (cur_time - m_TimeSinceLastPositionCheck), speed);
+					if (IsShadowStepExempted())
+					{
+						if (m_DistanceSinceLastPositionCheck > 800)
+						{
+							CheatDetected(MQWarpShadowStep, GetX(), GetY(), GetZ());
+						}
+					}
+					else if (IsKnockBackExempted())
+					{
+						//still potential to trigger this if you're knocked back off a
+						//HUGE fall that takes > 2.5 seconds
+						if (speed > 30.0f)
+						{
+							CheatDetected(MQWarpKnockBack, GetX(), GetY(), GetZ());
+						}
+					}
+					else if (!IsPortExempted())
+					{
+						if (!IsMQExemptedArea(zone->GetZoneID(), GetX(), GetY(), GetZ()))
+						{
+							if (speed > (runs * 2 * RuleR(Zone, MQWarpDetectionDistanceFactor)))
+							{
+								m_TimeSinceLastPositionCheck = cur_time;
+								m_DistanceSinceLastPositionCheck = 0.0f;
+								CheatDetected(MQWarp, GetX(), GetY(), GetZ());
+							}
+							else
+							{
+								CheatDetected(MQWarpLight, GetX(), GetY(), GetZ());
+							}
+						}
+					}
+				}
+			}
+		}
+		m_TimeSinceLastPositionCheck = cur_time;
+		m_DistanceSinceLastPositionCheck = 0.0f;
+	}
+	m_PortExemption = v;
+}
+
+const bool Client::IsMQExemptedArea(uint32 zoneID, float x, float y, float z) const
+{
+	float max_dist = 90000;
+	switch (zoneID)
+	{
+	case 2:
+	{
+		float delta = (x - (-713.6));
```

Comment:
Missing `f` suffix here and below for the all the float constants  `-713.6f` (Will avoid some double to float truncation warnings at least in msvc)